skrub/_multi_agg_joiner.py

```diff
@@ -0,0 +1,440 @@
+
+    suffixes : iterable of str, default=None
+        Suffixes to append to the `aux_tables`' column names.
+        If set to `None`, the table indexes in `aux_tables` are used,
+        e.g. for an aggregation of 2 `aux_tables`, "_1" and "_2" would be appended
+        to column names.
+
+    See Also
+    --------
+    AggJoiner :
+        Aggregate an auxiliary dataframe before joining it on a base dataframe.
+
+    Examples
+    --------
+    >>> import pandas as pd
+    >>> from skrub import MultiAggJoiner
+    >>> patients = pd.DataFrame({
+    ...    "patient_id": [1, 2],
+    ...    "age": ["72", "45"],
+    ... })
+    >>> hospitalizations = pd.DataFrame({
+    ...    "visit_id": range(1, 7),
+    ...    "patient_id": [1, 1, 1, 1, 2, 2],
+    ...    "days_of_stay": [2, 4, 1, 1, 3, 12],
+    ...    "hospital": ["Cochin", "Bichat", "Cochin", "Necker", "Bichat", "Bichat"],
+    ... })
+    >>> medications = pd.DataFrame({
+    ...    "medication_id": range(1, 6),
+    ...    "patient_id": [1, 1, 1, 1, 2],
+    ...    "medication": ["ozempic", "ozempic", "electrolytes", "ozempic", "morphine"],
+    ... })
+    >>> glucose = pd.DataFrame({
+    ...    "biology_id": range(1, 7),
+    ...    "patientID": [1, 1, 1, 1, 2, 2],
+    ...    "value": [1.4, 3.4, 1.0, 0.8, 3.1, 6.5],
+    ... })
+    >>> multi_agg_joiner = MultiAggJoiner(
+    ...    aux_tables=[hospitalizations, medications, glucose],
+    ...    main_keys=[["patient_id"], ["patient_id"], ["patient_id"]],
+    ...    aux_keys=[["patient_id"], ["patient_id"], ["patientID"]],
+    ...    cols=[["days_of_stay"], ["medication"], ["value"]],
+    ...    operations=[["max"], ["mode"], ["mean", "std"]],
+    ...    suffixes=["", "", "_glucose"],
+    ... )
+    >>> multi_agg_joiner.fit_transform(patients)
+       patient_id  age  ...  value_mean_glucose  value_std_glucose
+    0           1   72  ...                1.65           1.193035
+    1           2   45  ...                4.80           2.404163
+    """
+
+    def __init__(
+        self,
+        aux_tables,
+        *,
+        keys=None,
+        main_keys=None,
+        aux_keys=None,
+        cols=None,
+        operations=None,
+        suffixes=None,
+    ):
+        self.aux_tables = aux_tables
+        self.keys = keys
+        self.main_keys = main_keys
+        self.aux_keys = aux_keys
+        self.cols = cols
+        self.operations = operations
+        self.suffixes = suffixes
+
+    def _check_dataframes(self, X, aux_tables):
+        """Check dataframes input types.
+
+        Parameters
+        ----------
+        X : DataframeLike
+            The main table to augment.
+        aux_tables : iterator of DataframeLike or "X"
+            The auxiliary tables.
+
+        Raises
+        ------
+        TypeError
+            Raises an error if all the frames don't have the same type,
+            or if there is a Polars lazyframe.
+        """
+        if not _is_array_like(aux_tables):
```

Comment:
for this check and the other ones such as missing columns, couldn't we let the individual joiners do their checks to avoid duplicating the logic here?

Reply:
I agree that we could, the idea was that `MultiAggJoiner` is potentially going to be used to merge a large number of tables, possibly with many operations. I thought it was better to perform checks here directly rather than waiting the fits of the individual joiners

If you don't think it's going to be an issue I can remove them from here. It's true that this estimator becomes much more streamlined with letting some checks be performed in the individual joiners

Reply:
that's a good point. still I'd like to avoid duplication as much as possible to help us maintain consistent behavior between the joiner and multi joiner wherever it makes sense.

maybe the individual joiners could separate their checks from their fit and expose a `check_inputs` method to the multi joiner? it could then ask each joiner to perform checks before actually investing time to perform the joins, but the single joiner would keep the responsibility of saying whether it's able to perform a join or not

also, do we want to allow using a column from the first aux table to join the second aux table? if yes, it may not be possible to perform all checks ahead of time. either way we should state explicitly in the documentation whether that's allowed or not

Reply:
It's a good point. I'm rewriting the code so that individual joiners expose a `check_inputs` method, that I use in the multi joiner before fitting

For the second point it's a good question, for now it's not possible but could be explored for a later PR. I can open an issue on this point, as it's probably a case that we'll encounter in other estimators too

Reply:
Also, I think it makes sense to have some checks in the `MultiAggJoiner` too because the parameter names and input types are different, and this will not reflect if we mostly keep checks in the individual joiners

Reply:
> , for now it's not possible but could be explored for a later PR

that sounds good, let's state explicitly in the docstring that it is not (yet) supported

> Also, I think it makes sense to have some checks in the MultiAggJoiner too because the parameter names and input types are different, and this will not reflect if we mostly keep checks in the individual joiners

yes, the checks that are the same for both can be done in the individual joiners or in a module-level function, and the multijoiner can add other checks that are specific to the multi- case